### ISSUE TITLE
Dynamic ROC curve generation after segmentation training process

### ISSUE THREAD

**Issue by Cestovatels** [ultralytics/ultralytics#20098]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

After the segmentation training process, I want to create a ROC curve from the test data. Since I will get an automatic structure, I made a structure like the one below. However, I am not very sure about its accuracy. I would like to know if there is a different method. Thanks in advance for your help. @Y-T-G 

```py
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc
from pathlib import Path
import torch
import cv2
import glob
from ultralytics import YOLO
from PIL import Image, ImageDraw
import random

def polygon_to_mask(polygon_points, img_size):
    width, height = img_size
    polygon_points_scaled = [
        (int(polygon_points[i] * width), int(polygon_points[i+1] * height))
        for i in range(0, len(polygon_points), 2)
    ]
    mask = Image.new('L', (width, height), 0)
    draw = ImageDraw.Draw(mask)
    draw.polygon(polygon_points_scaled, outline=1, fill=1)
    return np.array(mask)

def load_txt_annotations(txt_path, img_size):
    class_masks = {}
    width, height = img_size
    if not Path(txt_path).exists():
        return class_masks
    with open(txt_path, 'r') as f:
        lines = f.readlines()
    for line in lines:
        parts = line.strip().split()
        class_id = int(parts[0])
        polygon_points = list(map(float, parts[1:]))
        mask = polygon_to_mask(polygon_points, (width, height))
        if class_id in class_masks:
            class_masks[class_id] = np.logical_or(class_masks[class_id], mask).astype(np.uint8)
        else:
            class_masks[class_id] = mask
    return class_masks

def sample_flattened_pixels(gt_mask, pred_mask, num_samples=2000):

    h, w = gt_mask.shape
    indices = random.sample(range(h * w), min(num_samples, h * w))
    gt_sampled = gt_mask.flatten()[indices]
    pred_sampled = pred_mask.flatten()[indices]
    return gt_sampled, pred_sampled

def plot_roc_curves_streaming(model, test_images_path, labels_path, class_names, max_images=1500):
    y_true_per_class = {i: [] for i in range(len(class_names))}
    y_pred_per_class = {i: [] for i in range(len(class_names))}
    
    test_images = sorted(glob.glob(f"{test_images_path}/*.jpg") + glob.glob(f"{test_images_path}/*.png"))[:max_images]
    
    for img_path in test_images:
        img_name = Path(img_path).stem
        img = cv2.imread(img_path)
        if img is None:
            continue
        height, width = img.shape[:2]
        img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        txt_path = f"{labels_path}/{img_name}.txt"
        gt_masks = load_txt_annotations(txt_path, (width, height))

        results = model(img_rgb)
        class_pred_masks = {}

        if results and hasattr(results[0], 'masks') and results[0].masks is not None:
            if len(results[0].masks) > 0:
                pred_masks = results[0].masks.data.cpu().numpy()
                pred_cls = results[0].boxes.cls.cpu().numpy().astype(int)
                pred_conf = results[0].boxes.conf.cpu().numpy()

                for i, cls in enumerate(pred_cls):
                    mask = pred_masks[i]
                    conf = pred_conf[i]
                    resized = cv2.resize(mask, (width, height)) * conf
                    if cls in class_pred_masks:
                        class_pred_masks[cls] = np.maximum(class_pred_masks[cls], resized)
                    else:
                        class_pred_masks[cls] = resized

        for class_idx in range(len(class_names)):
            gt_mask = gt_masks.get(class_idx, np.zeros((height, width), dtype=np.uint8))
            pred_mask = class_pred_masks.get(class_idx, np.zeros((height, width), dtype=np.float32))

            y_true_sampled, y_pred_sampled = sample_flattened_pixels(gt_mask, pred_mask, num_samples=2000)
            y_true_per_class[class_idx].extend(y_true_sampled)
            y_pred_per_class[class_idx].extend(y_pred_sampled)

    plt.figure(figsize=(10, 8))
    for class_idx, class_name in enumerate(class_names):
        y_true = np.array(y_true_per_class[class_idx])
        y_pred = np.array(y_pred_per_class[class_idx])

        if np.sum(y_true) < 10 or len(np.unique(y_true)) < 2:
            continue

        fpr, tpr, _ = roc_curve(y_true, y_pred)
        roc_auc = auc(fpr, tpr)
        plt.plot(fpr, tpr, lw=2, label=f'{class_name} (AUC = {roc_auc:.2f})')

    plt.plot([0, 1], [0, 1], 'k--', lw=2)
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('Streaming ROC - YOLOv8 Segment')
    plt.legend(loc="lower right")
    plt.grid(True, linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.savefig('streaming_roc_yolov8.png', dpi=300)
    plt.show()


model = YOLO('weights/best.pt')
model.cuda(1)

class_names = ["class1","class2","class3", ....]

plot_roc_curves_streaming(
    model=model,
    test_images_path='test/images',
    labels_path='test/labels',
    class_names=class_names
)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Cestovatels, thank you for sharing your detailed query and code example! 🚀 We appreciate your interest in using Ultralytics.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance on best practices and examples. Specifically, for segmentation-related tasks and post-training evaluations like generating ROC curves, you may find useful insights in the [YOLO Segmentation Guide](https://docs.ultralytics.com/tasks/segment/).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it effectively. For any ❓ Questions about custom implementations or training processes, please include as much information as possible, such as your dataset structure, training settings, and any relevant logs.

For further collaboration, you can join the Ultralytics community:
- Real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth conversations on [Discourse](https://community.ultralytics.com/)
- Share or explore knowledge on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Ensure you are using the latest version of the `ultralytics` package to avoid any outdated functionality. Upgrade with the following command:

```bash
pip install -U ultralytics
```

Check that your environment meets the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), including [Python>=3.8](https://www.python.org/) and [PyTorch>=1.8](https://pytorch.org/get-started/locally/).

## Environments

YOLO can be run in any of the following verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If the CI badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. These tests verify the functionality of YOLO across all [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu systems.

⚡ This is an automated response, but an Ultralytics engineer will review your issue and provide additional support shortly. Let us know if you need anything else! 😊
